### PR TITLE
Try PCA with masked data.

### PR DESCRIPTION
**Pull request by janewliang**
Masking does not aid matters, other than reducing the voxel by time matrix to ~50000 voxels instead of ~130000 or whatever it was. The first image is the proportion of variance explained among the first 20 principal components for Subject 10, no masking. The second is the same, but with masking. They're both sad, so I might move this discussion to an appendix or something, but I think we should scrap the idea altogether for our main analysis. 

@jarrodmillman and @matthew-brett : If it's of interest, the code to run PCA and generate these images is code/utils/scripts/pca_script.py, which should be the only modified file in this pull request. It looks like we won't actually be moving forward with this approach in our analysis, so it's not terribly important to have it all sorted out. I am indeed working with the time by time matrix after doing data_2d.T.dot(data_2d), so that isn't the problem. I'm using PCA from the sklearn.decomposition module to get just the first 20 principal components-- trying to do the entire decomposition (on my computer, at least) runs into nasty errors, as I mentioned today. 

pcasub010
maskedpcasub010

### CODE REVIEW

**Comment by matthew-brett** [comment]
Ah - I thought you were using the by-hand principal components from http://www.jarrodmillman.com/rcsds/lectures/pca_introduction.html#finding-the-principal-components-with-svd - but you are using sklearn PCA.

Here's me doing a by-hand PCA on the example image from class:

```
In [1]: import numpy as np

In [2]: import nibabel as nib

In [3]: img = nib.load('ds114_sub009_t2r1.nii')

In [4]: data = img.get_data()

In [5]: n_trs = data.shape[-1]

In [6]: time_by_vox = data.reshape((-1, n_trs)).T

In [7]: time_by_vox.shape
Out[7]: (173, 122880)

In [8]: t_by_v_centered = time_by_vox - np.mean(time_by_vox, axis=1)[:, None]

In [9]: unscaled_cov = t_by_v_centered.dot(t_by_v_centered.T)

In [10]: U, S, V = np.linalg.svd(unscaled_cov)

In [11]: import matplotlib.pyplot as plt

In [12]: plt.plot(np.sqrt(S))
Out[12]: [<matplotlib.lines.Line2D at 0x119bb5710>]

In [13]: plt.show()
```

Now I've played with your script, I see the problem, and it isn't your problem, it's numpy / nibabel.  Your data, when loaded, is of integer type, and it turns out `dot` is horribly slow on integers, and, for your integers, you are getting horrible horrible overflow errors.  Try changing line 40 of your script from:

```
    data = img.get_data()
```

to 

```
    data = img.get_data().astype(float)
```

What do you get after that?

By the way, generally I still prefer the long-hand PCA way, because it's easier for me to see what's going on.

In general, I'd always convert your data to floating point when you load it, to avoid problems like this.  It's my fault that I didn't make this clear during class.

**Comment by janewliang** [comment]
@matthew-brett Thanks for your comments! That helps clear things up. I had actually originally coded a PCA function by hand similar to what you did above/in the class notes, but got rid of it and started using sklearn's module when it was impossible to do the full decomposition.

I tried out converting the data to floats, and yes, it does get rid of the memory errors, speeds things up, and makes it possible to get all the principal components using either SVD by hand or the sklearn module. The resulting skree plots from either method look the same, but unfortunately also look completely different from the skree plots from before converting to floats. Nearly all the variance is captured in the first component, and all the other components capture a near-zero proportion of the variance, which feels quite suspicious. 

I checked the data values before and after converting to floats, and they are the same. Other than that, I'm out of ideas on why there's a difference in results because of the data type. Any insights?

floatpcasub010

**Comment by matthew-brett** [comment]
Yes, I'm afraid the data type was really badly messing you up.  What is happening, is that when doing `dot`, numpy is preserving your datatype.  The datatype is `np.int16` I think.   Let's imagine we're doing:

```
d = data_2d
c = d.T.dot(d)
```

`c[0, 0]` comes from `d[0, 0] ^ 2 + d[1, 0] ^ 2 + ... d[N-1, 0] ^ 2`.  If this value goes over the largest value that the int16 can hold (2^15-1 = 32767) then it will overflow to some large negative number:

```
In [8]: np.int16(32767) + np.int16(1)
/Users/mb312/.virtualenvs/rcsds/bin/ipython:1: RuntimeWarning: overflow encountered in short_scalars
  #!/Users/mb312/.virtualenvs/rcsds/bin/python
Out[8]: -32768

In [9]: np.int16(32767) + np.int16(2)
/Users/mb312/.virtualenvs/rcsds/bin/ipython:1: RuntimeWarning: overflow encountered in short_scalars
  #!/Users/mb312/.virtualenvs/rcsds/bin/python
Out[9]: -32767
```

So, what you were getting out of `dot` was basically random int16 values, depending on how many times the sum went round the the overflow cycle:

```
In [11]: arr = np.zeros((2, 3), dtype=np.int16) + np.int16(3201)

In [12]: arr.T.dot(arr)
Out[12]: 
array([[-19966, -19966, -19966],
       [-19966, -19966, -19966],
       [-19966, -19966, -19966]], dtype=int16)
```

It's unfortunate that numpy doesn't spit out a warning for this - I've made a numpy issue in case that can be fixed : https://github.com/numpy/numpy/issues/6753 . 

Did you subtract the mean from the columns of `data_2d` before doing `data_2d.T.dot(data_2d)`, and passing to the SVD / PCA?  Otherwise the first component will be dominated by this column mean.

I updated the PCA page to explain the relationship of the PCA, the SVD, and the SVD of the covariance : http://www.jarrodmillman.com/rcsds/lectures/pca_introduction.html#finding-the-principal-components-with-svd - let me know if you have any questions.

**Comment by janewliang** [comment]
Thanks for the clarifications. I wasn't subtracting the column means before, but I am now. It still results in the first component capturing almost all the variance, for either the masked or unmasked data.

**Comment by matthew-brett** [comment]
OK - I will take a look later today.

**Comment by matthew-brett** [comment]
Yes, I see the same thing with (`sub010` example):

```
from os.path import dirname, join as pjoin
import numpy as np
import numpy.linalg as npl

import matplotlib.pyplot as plt
plt.rcParams['image.interpolation'] = 'nearest'
plt.rcParams['image.cmap'] = 'gray'

from sklearn.decomposition import PCA

import nibabel as nib

DATA_PATH = pjoin(dirname(__file__), 'data', 'ds009')
SUB010_BOLD = pjoin(DATA_PATH,
                    'sub010', 'BOLD', 'task001_run001', 'bold.nii.gz')

img = nib.load(SUB010_BOLD)
# Cast data to float
data = img.get_data().astype(float)
# Drop dummy scans
data = data[..., 5:]
# Reshape to 2D
vol_shape, n_trs = data.shape[:-1], data.shape[-1]
data_2d = np.reshape(data, (-1, n_trs))
# Remove mean from voxels (columns)
data_2d = data_2d - np.mean(data_2d, axis=0)
# Uncomment below to remove mean over time (rows) using broadcasting
# data_2d = data_2d - np.mean(data_2d, axis=1)[:, None]
# Do PCA long-hand
cov = data_2d.T.dot(data_2d)
U, S, VT = npl.svd(cov)
# Variance proportion
S_prop = S / np.sum(S)
# Project data onto components
projections = data_2d.dot(U)
# Make images from components
proj_4d = np.reshape(projections, (vol_shape +  (n_trs,)))
# Save to disk
img = nib.Nifti1Image(proj_4d, img.affine, img.header)
nib.save(img, 'projections.nii')
```

There is a very strong effect of the mean over time in your data, that dominates the other effects.  Try uncommenting the line in the script to remove the mean over time (you would in any case remove the mean by the subsequent modeling).  Then you get a much more polite scree plot.  There are lots of movement effects in this dataset at least - I see from looking at the rest of the projection images.

**Comment by janewliang** [comment]
Plots make sense now (green is masked, blue is not masked)! Many thanks for the subtracting means solution, @matthew-brett.

maskedpcasub010

**Comment by benjaminleroy** [comment]
Looks good, I'm glad we worked out all the potential problems
